application/org.openjdk.jmc.flightrecorder.flameview/src/main/java/org/openjdk/jmc/flightrecorder/flameview/views/FlameGraphView.java

```diff
@@ -331,59 +338,86 @@ public void saveState(IMemento memento) {
 	public void selectionChanged(IWorkbenchPart part, ISelection selection) {
 		if (selection instanceof IStructuredSelection) {
 			Object first = ((IStructuredSelection) selection).getFirstElement();
-			setItems(AdapterUtil.getAdapter(first, IItemCollection.class));
+			IItemCollection items = AdapterUtil.getAdapter(first, IItemCollection.class);
+			if (items == null) {
+				rebuildModel(ItemCollectionToolkit.build(Stream.empty()));
+			} else if (!items.equals(currentItems)) {
+				rebuildModel(items);
+			}
 		}
 	}
 
-	private void setItems(IItemCollection items) {
-		if (items != null) {
-			currentItems = items;
-			rebuildModel(items);
+	private void rebuildModel(IItemCollection items) {
+		// Release old model calculation before building a new
+		if (modelCalculationFuture != null) {
+			modelCalculationFuture.cancel(true);
+			if (modelCalculationFuture.isDone()) {
+				finishModelCalculationFuture(modelCalculationFuture);
+			}
 		}
+
+		Callable<Void> modelPreparerTask = getModelPreparerTask(items);
+		modelCalculationFuture = MODEL_EXECUTOR.submit(modelPreparerTask);
+		this.currentItems = items;
+		this.modelState = ModelState.INIT;
 	}
 
-	private void rebuildModel(IItemCollection items) {
-		// Release old model before building the new
-		if (currentModelCalculator != null) {
-			currentModelCalculator.cancel(true);
+	private void finishModelCalculationFuture(Future<Void> f) {
+		try {
+			f.get();
+		} catch (CancellationException t) {
+			//noop
+		} catch (InterruptedException | ExecutionException e) {
+			FlightRecorderUI.getDefault().getLogger().log(Level.SEVERE, "Flameview build has failed", e); //$NON-NLS-1$
 		}
-		currentModelCalculator = getModelPreparer(items, frameSeparator, true);
-		currentModelCalculator.thenAcceptAsync(this::setModel, DisplayToolkit.inDisplayThread())
-				.exceptionally(FlameGraphView::handleModelBuildException);
 	}
 
-	private CompletableFuture<TraceNode> getModelPreparer(
-		final IItemCollection items, final FrameSeparator separator, final boolean materializeSelectedBranches) {
-		return CompletableFuture.supplyAsync(() -> {
+	private Callable<Void> getModelPreparerTask(final IItemCollection items) {
+		return () -> {
+			this.modelState = ModelState.CALCULATION;
 			StacktraceModel model = new StacktraceModel(threadRootAtTop, frameSeparator, items);
+
 			TraceNode root = TraceTreeUtils.createRootWithDescription(items, model.getRootFork().getBranchCount());
-			return TraceTreeUtils.createTree(root, model);
-		}, MODEL_EXECUTOR);
+			TraceNode traceNode = TraceTreeUtils.createTree(root, model);
+			StringBuilder jsonModelBuilder = toJSonModel(traceNode);
+
+			this.modelState = ModelState.READY;
+			DisplayToolkit.inDisplayThread().execute(() -> this.setModel(items, jsonModelBuilder.toString()));
```

Comment:
Want to get the string first, outside of the lambda?

Reply:
done!